Fix behavior of 'Editable Children' toggle

SaracenOne opened the pull request:
Toggling 'editable_children' off in the SceneTree will now no longer result in nodes owned by the currently edited scene, but assigned as children of foreign nodes, from being accidentally lost, where they before would still be visible in the scene, but not viewable since their parents were made hidden. They will instead be moved to the root of the instance, with their global transform perserved in the case of Node2D and Node3D derivatives. The editable children toggle will now also be subject to functional undoing and redoing.

Closes #46726

SaracenOne committed:
Fix behavior of 'Editable Children' toggle.

Prevents losing nodes owned by the edited scene when
toggling editable_children off on an instanced scene,
and makes the toggle compatible with undo-redo.

lyuma commented on editor/scene_tree_dock.cpp:
```diff
@@ -2012,17 +2012,169 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+			List<Node *> owned;
+			node->get_owned_by(p_owner, &owned, true);
+			// For Node2D and Node3D types, we can preserve the global transform when move them to the root node
+			if (typeid(RootType) == typeid(Node3D)) {
```
typeid() may introduce RTTI.

I'd suggest a compile-time check, or using a different function for each of the three cases.

akien-mga replied:
You should be able to do `Object::cast_to<Node3D>(p_root)` to check if it's Node3D.

SaracenOne replied:
It's not a huge deal, more of an attempt at good practice, but I think what I was hoping for was a means of using templates based on the root node to reduce unnessecary branching. Since my conversation with Lyuma that we determined that the way I was doing it would likely introduce RTTI, I shall attempt to change the way its done

KoBeWi commented:
#36301 would make the implementation simpler.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2038,17 +2038,155 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+			List<Node *> owned;
+			node->get_owned_by(p_owner, &owned, true);
+			// For Node2D and Node3D types, we can preserve the global transform when move them to the root node
+			if (Object::cast_to<Node3D>(p_root)) {
+				Node3D *node_3d = Object::cast_to<Node3D>(node);
+				if (node_3d) {
+					Transform3D gt = node_3d->get_global_transform();
+					node_3d->get_parent()->remove_child(node_3d);
+					p_root->add_child(node_3d);
+					node_3d->set_global_transform(gt);
+				} else {
+					node->get_parent()->remove_child(node);
+					p_root->add_child(node);
+				}
+			} else if (Object::cast_to<Node2D>(p_root)) {
+				Node2D *node_2d = Object::cast_to<Node2D>(node);
+				if (node_2d) {
+					Transform2D gt = node_2d->get_global_transform();
+					node_2d->get_parent()->remove_child(node_2d);
+					p_root->add_child(node_2d);
+					node_2d->set_global_transform(gt);
+				} else {
+					node->get_parent()->remove_child(node);
+					p_root->add_child(node);
+				}
+			} else {
+				node->get_parent()->remove_child(node);
+				p_root->add_child(node);
+			}
```
This can be replaced by the newly added `Node.reparent()`, which handles transforms.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2038,17 +2038,155 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		node->get_owned_by(p_owner, &owned, true);
+
+		Node3D *node_3d = Object::cast_to<Node3D>(node);
+		if (node_3d) {
+			Transform3D transform = p_transforms[i];
+			node_3d->get_parent()->remove_child(node_3d);
+			parent_node->add_child(node_3d);
+			node_3d->set_transform(transform);
+		} else {
+			Node2D *node_2d = Object::cast_to<Node2D>(node);
+			if (node_2d) {
+				Transform2D transform = p_transforms[i];
+				node_2d->get_parent()->remove_child(node_2d);
+				parent_node->add_child(node_2d);
+				node_2d->set_transform(transform);
+			} else {
+				node->get_parent()->remove_child(node);
+				parent_node->add_child(node);
+			}
+		}
```
Same.

SaracenOne commented:
Okay, this has been updated with the reparent function. Would recommend a new review since this bug is still relevant and still a bad UX trap.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
 }
 
+void SceneTreeDock::_update_all_gizmos(Node *p_node) {
+	Node3DEditor::get_singleton()->update_all_gizmos(p_node);
```
What is this for?

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+	List<Node *> nodes;
+	for (int i = 0; i < p_nodes.size(); i++) {
+		Node *node = Object::cast_to<Node>(p_nodes[i]);
+		ERR_FAIL_COND(node == nullptr);
```
```suggestion
		ERR_FAIL_NULL(node);
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+
+	for (int i = 0; i < p_nodes.size(); i++) {
+		Node *node = Object::cast_to<Node>(p_nodes[i]);
+		ERR_FAIL_COND(node == nullptr);
```
```suggestion
		ERR_FAIL_NULL(node);
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		ERR_FAIL_COND(node == nullptr);
+		NodePath np = p_paths[i];
+		Node *parent_node = p_root->get_node_or_null(np);
+		ERR_FAIL_COND(!parent_node);
```
```suggestion
		ERR_FAIL_NULL(parent_node);
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+	for (int i = 0; i < p_nodes.size(); i++) {
+		Node *node = Object::cast_to<Node>(p_nodes[i]);
+		ERR_FAIL_COND(node == nullptr);
+		NodePath np = p_paths[i];
```
```suggestion
		const NodePath &np = p_paths[i];
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		List<Node *> owned;
+		node->get_owned_by(p_owner, &owned, true);
+		node->reparent(parent_node);
+		node->set_name(p_names[i]);
```
You could store the previous name in a variable right before reparent. No need to keep it in Array I think.

SaracenOne replied:
Actually, I looked over this one, and no, the reason we keep it in an array is to revert naming conflicts. In the incident, that the node gets reparented to another node when its owner has its editable status removed, it may conflict with a matching name. By storing the names in an array, it means it can correctly restore its original name upon undoing.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+}
+
 void SceneTreeDock::_toggle_editable_children(Node *p_node) {
 	if (p_node) {
```
Since you are changing most of this function anyway, you can replace this block with an early return instead.
```suggestion
	if (!p_node) {
		return;
	}
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+			// Get the original paths, transforms, and names for undo
+			Array owned_nodes_array;
+			Array paths_array;
+			Array transform_array;
```
This array is unused.
Which makes me wonder whether separate method for transforms is neeeded.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+				}
+			}
+
+			undo_redo->add_undo_method(this, "_reparent_nodes_to_paths_with_transform_and_name", p_node, owned_nodes_array, paths_array, transform_array, name_array, edited_scene);
```
I think this method and the one below could be merged. You don't need transform and name arrays, as noted in comments above, and paths can be empty when reparenting to root. Though maybe I'm missing something.

SaracenOne replied:
While it seems overkill, I can explain why we have the names and transforms array. For explanation of names, see the explanation to the above response about undoing naming conflicts. As for the transforms, while its uncommon given a revert chain should put things back to its relative position before the reparent, storing the original transforms will restore it in situations where the local position might change. In example I was testing, I had a scene with an animated skeleton and bone attachments. If I instance this scene in another scene, make it editable, then attach a mesh to that bone attachment, then remove the editable status, the mesh will become detached. This is fine, but if the skeleton is animating, we don't know how to restore it local position by just using reparenting, hence the use of the transform array.

Admittedly, yeah, it this wasn't doing anything because the code to restore the local transform got lost during the migration to the reparent function, but I've restored it now.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+			bool original_scene_instance_load_placeholder = p_node->get_scene_instance_load_placeholder();
+
+			undo_redo->add_undo_method(p_node, "set_scene_instance_load_placeholder", original_scene_instance_load_placeholder);
+			undo_redo->add_do_method(p_node, "set_scene_instance_load_placeholder", false);
+		} else {
+			Node *edited_scene = EditorNode::get_singleton()->get_edited_scene();
+
+			List<Node *> owned;
+			p_node->get_owned_by(edited_scene, &owned, true);
+
+			// Get the original paths, transforms, and names for undo
+			Array owned_nodes_array;
+			Array paths_array;
+			Array transform_array;
+			Array name_array;
+
+			for (Node *owned_node : owned) {
+				owned_nodes_array.push_back(owned_node);
+				paths_array.push_back(p_node->get_path_to(owned_node->get_parent()));
+				name_array.push_back(owned_node->get_name());
+				Node3D *node_3d = Object::cast_to<Node3D>(owned_node);
+				if (node_3d) {
+					transform_array.push_back(node_3d->get_transform());
+				} else {
+					Node2D *node_2d = Object::cast_to<Node2D>(owned_node);
+					if (node_2d) {
+						transform_array.push_back(node_2d->get_transform());
+					} else {
+						transform_array.push_back(Variant());
+					}
+				}
+			}
+
+			undo_redo->add_undo_method(this, "_reparent_nodes_to_paths_with_transform_and_name", p_node, owned_nodes_array, paths_array, transform_array, name_array, edited_scene);
+			undo_redo->add_do_method(this, "_reparent_nodes_to_root", p_node, owned_nodes_array, edited_scene);
 		}
 
-		Node3DEditor::get_singleton()->update_all_gizmos(p_node);
+		undo_redo->add_undo_method(this, "_update_all_gizmos", p_node);
+		undo_redo->add_do_method(this, "_update_all_gizmos", p_node);
```
It would be simpler to bind "update_all_gizmos" than creating a new method.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -1831,13 +1831,16 @@ bool Node::is_greater_than(const Node *p_node) const {
 }
 
-void Node::get_owned_by(Node *p_by, List<Node *> *p_owned) {
+void Node::get_owned_by(Node *p_by, List<Node *> *p_owned, bool p_include_children_of_owned_by) {
```
Why did you change this method if you are always using the default anyway?

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2235,20 +2235,127 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		node->set_name(p_names[i]);
+		Node3D *node_3d = Object::cast_to<Node3D>(node);
+		if (node_3d) {
+			node_3d->set_transform(p_transforms[i]);
```
Does this do anything? `reparent()` already handles it.

SaracenOne commented:
Okay, I've made a bunch of updates and corrected most of what @KoBeWi pointed out. A lot of the issues were holdovers from the initial PR and I've corrected them now. There's a couple of points I've provided explanations for why its done that way though.

SaracenOne commented on editor/scene_tree_dock.cpp:
```diff
@@ -2224,17 +2224,111 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+	}
+
+	for (Node *node : nodes) {
+		node->set_owner(p_owner);
```
So I'm a little confused what you mean. The if statement is basically meant to check if the nodes in the array are parented to nodes owned by the instance we just removed editable children status from.

Admitedly though, after looking over it, I think this was the wrong place to put it. Instead, I've put it up in _toggle_editable_children when constructing the initial arrays, because otherwise, there's discontinuity between the two methods and this check doesn't reflect the description of what this specific method is meant to do.

KoBeWi replied:
What I meant was that `reparent()` clears the owner, so setting it before reparenting is pointless (you are still doing it). 

Previously your reparenting was conditional, so the unconditional `set_owner()` should go under else of that condition. This part is no longer relevant.

KoBeWi commented on editor/scene_tree_dock.h:
```diff
@@ -221,6 +221,11 @@ class SceneTreeDock : public VBoxContainer {
 
 	void _toggle_editable_children_from_selection();
+
+	void _update_all_gizmos(Node *p_node);
```
```suggestion
```

KoBeWi commented:
Something went wrong:

https://github.com/godotengine/godot/assets/2223172/cd3b0145-c56b-4423-9607-9384f253a153
```
Can't add child 'A' to itself.
Invalid owner. Owner must be an ancestor in the tree
```

SaracenOne commented:
Okay, sorry, I think I've fixed this *properly* this time.
- Fixed main node disappearing when removing editable children
- Made it so that update_all_gizmos method is binded properly and put it back seperately in the UndoRedo action to prevent the editable children status not correctly updating.
- Removed the empty method defintion.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2281,18 +2281,121 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		}
 
-		scene_tree->update_tree();
+		if (owned_nodes_array.size()) {
```
```suggestion
		if (!owned_nodes_array.is_empty()) {
```

KoBeWi reviewed (approved):
Works fine.

SaracenOne commented:
Changed PR and commit title to better match preferred submission format.

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2281,18 +2281,121 @@ void SceneTreeDock::_toggle_placeholder_from_selection() {
+		List<Node *> owned;
+		p_node->get_owned_by(edited_scene, &owned);
+
+		// Get the original paths, transforms, and names for undo
```
```suggestion
		// Get the original paths, transforms, and names for undo.
```

akien-mga commented:
Rebased and amended to apply @AThousandShips' suggestions.

akien-mga commented:
Thanks!

bhottovy commented:
After this was merged I have started getting errors like:
`scene/resources/packed_scene.cpp:1648 - Index p_idx = 6 is out of bounds (nodes.size() = 3).`
spammed in the editor output whenever I save a scene containing an instanced child scene that is inherited and 'editable_children' is enabled. If 'editable_children' is disabled I don't have any issue.

See #87424
